### ISSUE TITLE
Test 80-test_cmp_http hangs on NonStop x86 when run without a valid tty

### ISSUE THREAD

**Issue by rsbeckerca** [openssl/openssl#15781]
I'm not sure what is happening here, but when I run the test through our Jenkins CI test, which is not currently verbose, the test hangs 100% of the time - the terminal is defunct when this happens. When run from an interactive shell, whether verbose or not, the test passes.

This occurs at commit 835dd706d. It has been happening since the test was known to pass on the platform at alpha17. Unfortunately, I could not throw the hung process into debug - that may change with verbose on on the next run.

I am going to be rerunning the test cycle with verbose turned on to see whether I can find which subtest is having issues. The stack trace in openssl is as follows, which seems to be indicating that our friend problem server port is not being communicated between the openssl (at `BIO_accept_ex`) and the test script:
```
Procedure call trace:
  AsmBreakDispatchNoRelink + 0x2 (Milli)
  $n_NSK_Event_Wait + 0xCA (Milli)
  Timer_WaitOnEmbeddedTLE_ + 0xDE (SLr)
  PROCESS_WAIT_ + 0x38 (SLr)
  WAIT + 0x36 (SLr)
  tsleep_l + 0x359 (DLL zcipdll)
  csif_oss_accept64 + 0x20E5 (DLL zcipdll)
  QIOCIP_OSS_FS_ACCEPT_ + 0x1FD (SLr)
  Pxso_SocketAccept_Net_ + 0x107A (SLr)
  PXL2_SOCKET_ACCEPT_ + 0x411 (SLr)
  accept + 0x597 (SLr)
  $n_EnterPriv + 0x540 (Milli)
  BIO_accept_ex + 0x4F (DLL libcrypto.so)
  acpt_state + 0x60E (DLL libcrypto.so)
  acpt_ctrl + 0xD8 (DLL libcrypto.so)
  BIO_ctrl + 0xCC (DLL libcrypto.so)
  http_server_get_asn1_req + 0xAC (UCr)
  cmp_server + 0x12F (UCr)
  cmp_main + 0xC17 (UCr)
  do_cmd + 0x82 (UCr)
  main + 0x2A0 (UCr)
  _MAIN + 0x3F (UCr)
```
The situation is similar to running the test from inside a Docker container.

**Comment by rsbeckerca** [comment]
The test suite ran until here:
```
ok
80-test_cipherlist.t ............... 
# The results of this test will end up in test-runs/test_cipherlist
../../util/wrap.pl ../../test/versions => 0
1..1
    # Subtest: ../../test/cipherlist_test
    1..3
    ok 1 - test_default_cipherlist_implicit
    ok 2 - test_default_cipherlist_explicit
    ok 3 - test_default_cipherlist_clear
../../util/wrap.pl ../../test/cipherlist_test => 0
ok 1 - running cipherlist_test
ok
80-test_ciphername.t ............... 
# The results of this test will end up in test-runs/test_ciphername
1..1
    # Subtest: ../../test/ciphername_test
    1..1
    ok 1 - test_cipher_name
../../util/wrap.pl ../../test/ciphername_test => 0
ok 1 - running ciphername_test
ok
```
Then hung.

**Comment by rsbeckerca** [comment]
There was a lot in the Jenkins I/O buffer:
```
# The results of this test will end up in test-runs/test_cmp_http
1..5
Current directory is /home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/test/recipes/80-test_cmp_http_data/Mock
Launching mock server: ../../../../util/wrap.pl ../../../../apps/openssl cmp -config server.cnf
Pid is: 1795162217
Server output: cmp_main:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2674:CMP info: using section(s) 'cmp' of OpenSSL configuration file 'server.cnf'
Server output: ACCEPT [::]:61380 PID=1040187525
# Subtest: CMP app CLI Mock connection
    1..37
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
cmp: Received request, 1st line: POST /pkix/ HTTP/1.0
cmp: Received request, 1st line: POST /pkix/ HTTP/1.0
# cmp_main:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2674:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2197:CMP warning: -proxy option argument is empty string, resetting option
# setup_client_ctx:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:1897:CMP info: will contact http://127.0.0.1:61380/pkix/
# CMP info: sending IR
# CMP info: received IP
# CMP info: sending CERTCONF
# CMP info: received PKICONF
# save_free_certs:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:1947:CMP info: received 1 enrolled certificate(s), saving to file '../../../../test-runs/test_cmp_http/test.cert.pem'
../../../../util/wrap.pl ../../../../apps/openssl cmp -server '127.0.0.1:61380' -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 => 0
    ok 1 - default config
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
# cmp_main:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2674:CMP info: using section(s) 'Mock connection' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2197:CMP warning: -proxy option argument is empty string, resetting option
# setup_client_ctx:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:1897:CMP info: will contact http://xn--rksmrgs-5wao1o.example.com:61380/pkix/
# CMP info: sending IR
# CMP error: connect timeout
# CMP error: transfer error:request sent: IR, expected response: IP
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -server 'xn--rksmrgs-5wao1o.example.com:61380' -msg_timeout 1 => 255
    ok 2 - wrong server

<BIG SNIP>

    ok 88 - kur empty oldcert file
Warning: certificate from 'trusted.crt' with subject '/O=openssl_cmp' is not a CA cert
setup_protection_ctx:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:1344:CMP error: must give both -cert and -key options or neither
cmp_main:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2780:CMP error: cannot set up CMP context
# cmp_main:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2674:CMP info: using section(s) 'Mock enrollment' of OpenSSL configuration file '../Mock/test.cnf'
# opt_str:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2197:CMP warning: -proxy option argument is empty string, resetting option
# opt_str:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2197:CMP warning: -cert option argument is empty string, resetting option
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock enrollment' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -cmd kur -newkey new.key -newkeypass 'pass:' -certout ../../../../test-runs/test_cmp_http/test.certout_kur8.pem -out_trusted root.crt -cert "" -server '127.0.0.1:61380' => 255
    ok 89 - kur without cert and oldcert
ok 5 - CMP app CLI Mock enrollment
# 
Killing mock server with pid=1040187525
ok
```
The hang actually occurred here. I stopped the OpenSSL process that had `/test/recipes/80-test_cmp_http_data/Mock` open, got the flush to occur, and the test continued with only this test hanging, so I'm suspecting we're back to a perl->openssl mis-communication in the test based on being hung at the `accept`.

**Comment by rsbeckerca** [comment]
I'm going to retest on a more recent version of the operating systems. This might impact some `fcntl()` calls.

**Comment by levitte** [comment]
I doubt very much that it's caused by 835dd70
*However*, there was a change in the PID handling in #15642.  That might be significant...

**Comment by rsbeckerca** [comment]
> 
> 
> I doubt very much that it's caused by [835dd70](https://github.com/openssl/openssl/commit/835dd706d3a916dacdb302905899a32638ed8adc)
> _However_, there was a change in the PID handling in #15642. That might be significant...

@levitte I hope it helps. Still hangs at c3932c34c. If there's anything I can do to help track this down, please let me know.

**Comment by levitte** [comment]
> If there's anything I can do to help track this down, please let me know.

Considering you're pretty alone with NonStop around here, and we've noticed NonStop specific things (such as non-standard behavior with some POSIX functions, if I recall correctly) happen before, I'm afraid it falls on you to do the first analysis.

**Comment by rsbeckerca** [comment]
> 
> 
> > If there's anything I can do to help track this down, please let me know.
> 
> Considering you're pretty alone with NonStop around here, and we've noticed NonStop specific things (such as non-standard behavior with some POSIX functions, if I recall correctly) happen before, I'm afraid it falls on you to do the first analysis.

I guess, something wrong is happening somewhere in here:
```
    ok 89 - kur without cert and oldcert
ok 5 - CMP app CLI Mock enrollment
# 
Killing mock server with pid=1040187525
ok```

But I don't even know where to start looking. The `Killing` bit happens after I stop the `openssl` server waiting on accept. Somewhere in here, openssl starts, serves up the port on which it is listening, but `perl` is not picking it up, or is waiting to receive it, or is trying to connect to the wrong server, which, with Jenkins running, could have a port `80`. I assume there is a `fflush()` in `openssl` after the port is reported, otherwise `perl` may not see it. Everything is heavily buffered in a Jenkins environment because `stdout` and `stderr` are both non-interactive ttys. I just don't really get where in the recipe to look on the `perl` side.

**Comment by t8m** [comment]
Could you somehow try to strace the processes created? Assuming you have strace on NonStop. Add strace to the execution of the processes in the test or in the wrap.pl script.

**Comment by mattcaswell** [comment]
I've just seen this same problem on Solaris. It hangs at this point:

````
../../../../util/wrap.pl ../../../../apps/openssl cmp -config ../Mock/test.cnf -section 'Mock enrollment' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 -cmd kur -newkey new.key -newkeypass 'pass:' -certout ../../../../test-runs/test_cmp_http/test.certout_kur8.pem -out_trusted root.crt -cert "" -server '127.0.0.1:52880' => 1
    ok 89 - kur without cert and oldcert
ok 5 - CMP app CLI Mock enrollment
# 
Killing mock server with pid=13705
````

**Comment by levitte** [comment]
> I assume there is a fflush() in `openssl` after the port is reported, otherwise `perl` may not see it.

This is not true.  From your own verbose output:

```
Server output: cmp_main:/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/apps/cmp.c:2674:CMP info: using section(s) 'cmp' of OpenSSL configuration file 'server.cnf'
Server output: ACCEPT [::]:61380 PID=1040187525
```
...
```
../../../../util/wrap.pl ../../../../apps/openssl cmp -server '127.0.0.1:61380' -config ../Mock/test.cnf -section 'Mock connection' -certout ../../../../test-runs/test_cmp_http/test.cert.pem -proxy '' -no_proxy 127.0.0.1 => 0
```
...
```
Killing mock server with pid=1040187525
```

If you look at the numbers, you can see that both the port and the PID were correctly picked up.

However, the mock server is opened with a pipe:

https://github.com/openssl/openssl/blob/98dc656e5f491cc29a61892d152c9779527da6f8/test/recipes/80-test_cmp_http.t#L274

and something to be noted is that the file handle `$server_fh` is never explicitly closed.  Perl will, however, attempt to close it (like any file handle that's still present) when ending `80-test_cmp_http.t`, and I'm guessing that this is where it hangs.  If you apply this patch, I think you'll see that confirmed:

``` diff
diff --git a/test/recipes/80-test_cmp_http.t b/test/recipes/80-test_cmp_http.t
index 10f2b84c68..7c19e142cd 100644
--- a/test/recipes/80-test_cmp_http.t
+++ b/test/recipes/80-test_cmp_http.t
@@ -295,4 +295,7 @@ sub stop_mock_server {
     my $pid = $_[0];
     print "Killing mock server with pid=$pid\n";
     kill('QUIT', $pid);
+    note('Trying to close $server_fh');
+    close $server_fh;
+    note('Successfully closed $server_fh');
 }
```

If I guessed correctly, you'll see the first note, but not the second.

**Comment by levitte** [comment]
Another thing to check, when the kill message has been reached, is whether the process has actually been killed.  The best way to do that, I think, is this (you might have to change the `ps` arguments):

```
ps auxw | grep 'openssl  *cmp'
```

See if any of the processes have a PID corresponding to what the killing line says

**Comment by mattcaswell** [comment]
Just tried your patch on solaris:

````
ok 5 - CMP app CLI Mock enrollment
# 
Killing mock server with pid=16768
# Trying to close $server_fh
````

And running your `ps` command I see:

````
matt@solarisvm:~$ ps -ef | grep 'openssl  *cmp'
    matt 16767 16766   0 09:35:04 pts/1       0:00 perl ../../../../util/wrap.pl ../../../../apps/openssl cmp -config server.cnf
    matt 16768 16767   1 09:35:04 pts/1       0:01 ../../../../apps/openssl cmp -config server.cnf
````

So, it seems you are right, it is hanging attempting to close the `$server_fh`, and the process has not in fact been killed.

**Comment by levitte** [comment]
You might want to try `ps -efl` to see in what state the process is

**Comment by levitte** [comment]
I suspect it's become a zombie, and is therefore still hanging on to its stdout but is otherwise unresponsive.  That would explain why closing hangs

**Comment by levitte** [comment]
What happens if you replace `kill('QUIT', $pid);` with `kill('KILL', $pid);`?  KILL is the infamous -9 signal...

**Comment by mattcaswell** [comment]
````
matt@solarisvm:~$ ps -efl | grep 16768
 0 S     matt 16768 16767   0  40 20        ?   3256        ? 09:35:04 pts/1       0:01 ../../../../apps/openssl cmp -confi
````
I'm not sure what this telss me.

**Comment by levitte** [comment]
The two first columns are process flags and process status codes.  From `man ps`:

```
PROCESS FLAGS
       The sum of these values is displayed in the "F" column, which is
       provided by the flags output specifier:

               1    forked but didn't exec
               4    used super-user privileges

PROCESS STATE CODES
       Here are the different values that the s, stat and state output
       specifiers (header "STAT" or "S") will display to describe the state of
       a process:

               D    uninterruptible sleep (usually IO)
               I    Idle kernel thread
               R    running or runnable (on run queue)
               S    interruptible sleep (waiting for an event to complete)
               T    stopped by job control signal
               t    stopped by debugger during the tracing
               W    paging (not valid since the 2.6.xx kernel)
               X    dead (should never be seen)
               Z    defunct ("zombie") process, terminated but not reaped by
                    its parent
```

So not a zombie, but still something that refused to die.  Please try the KILL change and see if that helps

**Comment by mattcaswell** [comment]
> What happens if you replace kill('QUIT', $pid); with kill('KILL', $pid);? KILL is the infamous -9 signal...

It works!! The hang is gone.

**Comment by levitte** [comment]
> It works!! The hang is gone.

Good.  It's a pretty brutal way to resolve the issue, but if it works, it works.

**Comment by mattcaswell** [comment]
Will you raise a PR for this?

**Comment by levitte** [comment]
Sure

**Comment by levitte** [comment]
@rsbeckerca, I hope #15797 fixes things for you as well.  We're hurrying the fix now because we want it in the beta1 release, which is happening today.

**Comment by rsbeckerca** [comment]
> @rsbeckerca, I hope #15797 fixes things for you as well.  We're hurrying that fix now because we want it in the beta1 release, which is happening today.

Will try first thing when I can drag myself out of bed this AM

**Comment by rsbeckerca** [comment]
Quick explanation on why this might be the correct fix on some platforms:

`kill('QUIT', $pid);` is considered a deferrable signal, meaning that it will not be delivered while the code is outside of user space (a.k.a. in the operating system, waiting for something to happen). `accept()` on NonStop, will not deliver `SIGQUIT` to the process until `accept()` completes.

`kill('KILL', $pid);` is a non-deferrable signal, meaning that it is delivered immediately. It also cannot be caught, so `SIG_IGN` cannot be set for it. This will immediately kill the process regardless of what state it is in (unless zombied, which is very rare on NonStop).

The behaviour above is consistent with how I have stopped the `openssl` process to resume tests after they hang, by using `kill -9 pid` instead of another signal, which did not work.

Aside: `strace` is not available, but I can obtain a snapshot of what the process condition is using a program called `pstate`, which gives me all file opens, current stack trace, DLL use, register state, run state, memory used, and whether a deferred signal has been received. I did not notice/look for this last condition before but will next time.

The build test for 38fc02a708 is currently in progress and will advise.
